luxonis_ml/data/datasets/luxonis_dataset.py

```diff
@@ -529,76 +370,116 @@ def _add_process_batch(batch_data: List[Annotation]) -> None:
-                    elif instance_id not in new_index["instance_id"]:
-                        new_index["instance_id"].append(instance_id)
+                            # TODO: we may also want to check for duplicate uuids to get a one-to-one relationship
+                    elif uuid not in new_index["uuid"]:
+                        new_index["uuid"].append(uuid)
                         new_index["file"].append(file)
-                        new_index["original_filepath"].append(osp.abspath(filepath))
+                        new_index["original_filepath"].append(str(filepath.absolute()))
 
-                    self.pfm.write(ann.to_parquet(instance_id))
-                    progress.update(task, advance=1)
-                progress.stop()
+                    self.pfm.write({"uuid": uuid, **ann.to_parquet()})
+                    self.progress.update(task, advance=1)
+                self.progress.stop()
 
         if self.bucket_storage == BucketStorage.LOCAL:
-            self.pfm = ParquetFileManager(self.annotations_path)
+            self.pfm = ParquetFileManager(str(self.annotations_path))
         else:
             self._make_temp_dir()
-            annotations_dir = osp.join(self.tmp_dir, "annotations")
-            os.makedirs(annotations_dir, exist_ok=True)
-            self.pfm = ParquetFileManager(annotations_dir)
+            annotations_dir = self.tmp_dir / "annotations"
+            annotations_dir.mkdir(exist_ok=True, parents=True)
+            self.pfm = ParquetFileManager(str(annotations_dir))
 
         index = self._get_file_index()
-        new_index = {"instance_id": [], "file": [], "original_filepath": []}
+        new_index = {"uuid": [], "file": [], "original_filepath": []}
 
-        batch_data: list[Annotation] = []
+        batch_data: list[DatasetRecord] = []
 
-        for i, data in enumerate(generator):
-            batch_data.append(Annotation.from_dict(data))
-            if (i + 1) % batch_size == 0:
+        classes_per_task: Dict[str, Set[str]] = defaultdict(set)
+        num_kpts_per_task: Dict[str, int] = {}
+
+        for i, data in enumerate(generator, start=1):
+            record = data if isinstance(data, DatasetRecord) else DatasetRecord(**data)
+            if record.annotation is not None:
+                classes_per_task[record.annotation.task].add(record.annotation.class_)
+                if record.annotation.type_ == "keypoints":
+                    num_kpts_per_task[record.annotation.task] = len(
+                        record.annotation.keypoints
+                    )
+
+            batch_data.append(record)
+            if i % batch_size == 0:
                 _add_process_batch(batch_data)
                 batch_data = []
 
         _add_process_batch(batch_data)
 
+        _, curr_classes = self.get_classes()
+        if not curr_classes:
+            for task, classes in classes_per_task.items():
+                self.set_classes(list(classes), task)
```

Comment:
Would this code work automatically if we call `add` many times on the same dataset? Or would it only detect changes for the first time?

Reply:
Yeah, it would only detect the classes for the first time. So if we'd ingest an additional class, the classes would have to be set manually. I can improve it so it detects a new class and adds it automatically too.

Reply:
> I can improve it so it detects a new class and adds it automatically too.

It would be nice-to-have but not strictly necessary for the PR IMO